brainiak/reprsimil/brsa.py

```diff
@@ -0,0 +1,1118 @@
+        assert X.ndim == 2, 'The data should be 2 dimension ndarray'
+
+        assert np.all(np.std(X, axis=0) > 0),\
+            'The time courses of some voxels do not change at all.'\
+            ' Please make sure all voxels are within the brain'
+
+        # check design matrix
+        assert_all_finite(design)
+        assert design.ndim == 2,\
+            'The design matrix should be 2 dimension ndarray'
+        assert np.all(np.std(design, axis=0) >= 0) or self.pad_DC,\
+            'You already included DC component in the '\
+            'design matrix. Please set pad_DC as False'
+        assert np.size(design, axis=0) == np.size(X, axis=0),\
+            'Design matrix and data do not '\
+            'have the same number of time points.'
+        if self.pad_DC:
+            logger.debug('Padding one more column of 1 to '
+                         'the end of design matrix.')
+            design = np.concatenate((design,
+                                     np.ones([design.shape[0], 1])), axis=1)
+        assert self.rank is None or self.rank <= design.shape[1],\
+            'Your design matrix has fewer columns than the rank you set'
+
+        # check scan_onsets validity
+        assert scan_onsets is None or\
+            (np.max(scan_onsets) <= X.shape[0] and np.min(scan_onsets) >= 0),\
+            'Some scan onsets provided are out of the range of time points.'
+
+        # check the size of coords and inten
+        if self.GP_space:
+            logger.debug('Fitting with Gaussian Process prior on log(SNR)')
+            assert coords is not None and coords.shape[0] == X.shape[1],\
+                'Spatial smoothness was requested by setting GP_space. '\
+                'But the voxel number of coords does not match that of '\
+                'data X, or voxel coordinates are not provided. '\
+                'Please make sure that coords is in the shape of '\
+                '[ n_voxel x 3].'
+            assert coords.ndim <= 2,\
+                'The coordinate matrix should not have '\
+                'more than 2 dimensions'
+            if coords.ndim == 1:
+                coords = np.expand_dims(coords, 1)
+            if self.GP_inten:
+                assert inten is not None and inten.shape[0] == X.shape[1],\
+                    'The voxel number of intensity does not '\
+                    'match that of data X, or intensity not provided.'
+                assert np.var(inten) > 0,\
+                    'All voxels have the same intensity.'
+                if inten_weight is None:
+                    inten_weight = self.inten_weight
+                assert inten_weight >= 0 and inten_weight <= 1,\
+                    'The weight of intensity kernel is '\
+                    'out of the range of (0,1).'\
+                    'I suggest using a small value, such as 0.1.'
+            else:
+                inten_weight = 0.0
+
+        # Run Bayesian RSA
+        if not self.GP_space:
```

Comment:
I think right now if I pass in onsets without asking for the GP prior then they get silently ignored. Would it be better to check above for incompatible combinations of parameters like this, and then just let `scan_onsets` get passed through here? Just having some trouble following the logic -- another thing that might help is just commenting on what the different dispatchers are (I think it's BRSA without GP spatial prior for the first one, then BRSA without GP prior on intensity, then BRSA with both).